src/codegen/machine_ir.c

```diff
@@ -0,0 +1,326 @@
+#include "codegen/codegen_forward.h"
+#include "codegen/machine_ir_forward.h"
+#include "utils.h"
+#include <codegen/machine_ir.h>
+
+#include <codegen.h>
+#include <codegen/intermediate_representation.h>
+
+MIROperand mir_op_function(IRFunction *f) {
+  MIROperand out = {0};
+  out.kind = MIR_OP_FUNCTION;
+  out.value.function = f;
+  return out;
+}
+
+MIROperand mir_op_block(IRBlock *block) {
+  MIROperand out = {0};
+  out.kind = MIR_OP_BLOCK;
+  out.value.block = block;
+  return out;
+}
+
+MIROperand mir_op_reference(MIRInstruction *m_inst, IRInstruction *inst) {
+  ASSERT(inst, "Invalid argument");
+  if (!inst->machine_inst) {
+    ir_femit_instruction(stdout, inst);
+    ICE("Must translate IRInstruction into MIR before taking reference to it.");
+  }
+  MIROperand out = {0};
+  out.kind = MIR_OP_REFERENCE;
+  out.value.ref = inst->machine_inst;
+  return out;
+}
+
+MIROperand mir_op_immediate(int64_t imm) {
+  MIROperand out = {0};
+  out.kind = MIR_OP_IMMEDIATE;
+  out.value.imm = imm;
+  return out;
+}
+
+MIROperand mir_op_name(const char *name) {
+  MIROperand out = {0};
+  out.kind = MIR_OP_NAME;
+  out.value.name = name;
+  return out;
+}
+
+MIROperand mir_op_register(RegisterDescriptor reg, uint16_t size) {
+  MIROperand out = {0};
+  out.kind = MIR_OP_REGISTER;
+  out.value.reg.value = reg;
+  out.value.reg.size = size;
+  return out;
+}
+
+static size_t mir_alloc_id = 0;
+MIRInstruction *mir_makenew(uint32_t opcode) {
+  MIRInstruction *mir = calloc(1, sizeof(*mir));
+  ASSERT(mir, "Memory allocation failure");
+  mir->id = ++mir_alloc_id;
```

Comment:
I'm not really arguing it's a bad idea to get rid of the statics here---that's definitely a good idea---but I do wonder what it would accomplish

Reply:
> I'm not really arguing it's a bad idea to get rid of the statics here---that's definitely a good idea---but I do wonder what it would accomplish

Nothing unless you’re planning to be able to codegen several programs/files concurrently. If you care about multithreading at all, then these should all go in the context because that’s what it’s for, otherwise none of that matters to begin with and everything that’s in the context could just as well be a bunch of globals. My point here is, it doesn’t make that much sense to put half of this stuff in the context and make the other half global variables. It’s either one way or the other, but mixing both unfortunately doesn’t end up accomplishing much of anything because you get both the inconvenience of having to pass a context to every function and no thread safety.